zebra-network/src/protocol/external/codec.rs

```diff
@@ -152,26 +147,54 @@ impl Encoder<Message> for Codec {
             FilterAdd { .. } => b"filteradd\0\0\0",
             FilterClear { .. } => b"filterclear\0",
         };
-        trace!(?item, len = body.len());
+        trace!(?item, len = body_length);
 
-        // XXX this should write directly into the buffer,
-        // but leave it for now until we fix the issue above.
-        let mut header = [0u8; HEADER_LEN];
-        let mut header_writer = Cursor::new(&mut header[..]);
-        header_writer.write_all(&Magic::from(self.builder.network).0[..])?;
-        header_writer.write_all(command)?;
-        header_writer.write_u32::<LittleEndian>(body.len() as u32)?;
-        header_writer.write_all(&sha256d::Checksum::from(&body[..]).0)?;
+        dst.reserve(HEADER_LEN + body_length);
+        let start_len = dst.len();
+        {
+            let dst = &mut dst.writer();
+            dst.write_all(&Magic::from(self.builder.network).0[..])?;
+            dst.write_all(command)?;
+            dst.write_u32::<LittleEndian>(body_length as u32)?;
+
+            // We zero the checksum at first, and compute it later
+            // after the body has been written.
+            dst.write_u32::<LittleEndian>(0)?;
 
-        dst.reserve(HEADER_LEN + body.len());
-        dst.extend_from_slice(&header);
-        dst.extend_from_slice(&body);
+            self.write_body(&item, dst)?;
+        }
+        let checksum = sha256d::Checksum::from(&dst[start_len + HEADER_LEN..]);
+        dst[start_len + 20..][..4].copy_from_slice(&checksum.0);
 
         Ok(())
     }
 }
 
 impl Codec {
+    /// Obtain the size of the body of a given message. This will match the
+    /// number of bytes written to the writer provided to `write_body` for the
+    /// same message.
+    fn body_length(&self, msg: &Message) -> usize {
```

Comment:
```suggestion
    ///
    /// TODO: Replace with a size estimate, to avoid multiple serializations
    /// for large data structures like lists, blocks, and transactions.
    fn body_length(&self, msg: &Message) -> usize {
```